1623 add training

hiker opened the pull request:
Adds the PSyclone training material (based on #3131)

hiker commented:
This PR affects a few files outside of training. These are mostly small/cosmetic changes, so I didn't bother taking them out:
1. `.github/workflow` - adding training to the pipelines
2. `doc/psyadd/.gitignore` - ignore build files
3. `examples/gocean/eg5/extract/Makefile` - remove unnecessary white space
4.` examples/lfric/eg17*/main.x90` - removed unnecessary imports
5. `.../driver_creator_test.py` - added an explicit test for creating valid unit names (the function is covered anyway, but it does not do any real work, while with this test it makes sure that replacing happens as expected)

arporter commented on examples/gocean/eg5/extract/Makefile:
```diff
@@ -123,7 +123,6 @@ compile: transform $(NAME) $(DRIVER_INIT).$(TYPE) $(DRIVER_UPDATE).$(TYPE)
 
 transform: psy.f90
 
-
```
Please revert this file.

hiker replied:
Oops - done.

arporter commented on tutorial/training/Makefile:
```diff
@@ -0,0 +1,53 @@
+
+
+test:
+	$(MAKE) -C users test
```
As commented earlier, "users" doesn't seem very intuitive here.

hiker replied:
Done.

arporter commented on tutorial/training/Makefile:
```diff
@@ -0,0 +1,53 @@
+	$(MAKE) -C lfric test
+	$(MAKE) -C transformation test
+
+compile:
```
Can you do:
```make
test compile clean:
    $(MAKE) -C gocean $@
    $(MAKE) -C lfric $@
    ...
```
?

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,129 @@
+done automatically by PSYclone.
+
+PSyclone relies on the meta-data provided for each kernel. Especially
+in `count_neighbours_mod` the declaration is:
```
Perhaps s/declaration/metadata/

hiker replied:
Done.

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,129 @@
+about empty and filled cells, is accessed using a stencil-operation
+of depth 1. The infrastructure library will do the required halo
+exchange, acquiring the current data from all eight neighbours.
+And PSyclone will automatically insert calls to halo-exchanges
```
s/. And/and/

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,129 @@
+             -oalg time_step_alg_mod.f90 -opsy time_step_alg_mod_psy.f90 time_step_alg_mod.x90
+
+
+Inspecting the PSy-layer output file `time_step_alg_mod_psy.f90` shows:
```
s/output file/file created by PSyclone,/

hiker replied:
Done.

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,129 @@
+the halo exchange is automatically triggered.
+
+> Note that at this stage the infrastructure library (and therefore also
+> PSyclone) only supports halo exchanges of depth 1.
```
Perhaps add a note that the LFRic DSL doesn't have this limitation?

hiker replied:
Added.

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/fuse_loops.py:
```diff
@@ -0,0 +1,85 @@
+
+    '''
+    fuse = GOceanLoopFuseTrans()
+    inline = KernelModuleInlineTrans()
```
Probably best to call this "mod_inline" or something.

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/openmp_combined.py:
```diff
@@ -0,0 +1,82 @@
+    omp_parallel = OMPParallelTrans()
+    # Optional argument: schedule
+    omp_do = OMPLoopTrans(omp_schedule="dynamic")
+    inline = KernelModuleInlineTrans()
```
"mod_inline" and then "Module inline..." below.

hiker replied:
Done.

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/psy.f90:
```diff
@@ -0,0 +1,106 @@
+  MODULE psy_time_step_alg_mod
```
What's the purpose of this file? I would have thought we just generated this with PSyclone?

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/solution/fuse_loops.py:
```diff
@@ -0,0 +1,85 @@
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Python script intended to be passed to PSyclone's generate()
+function via the -s option. It adds kernel fuseion code to
```
"It fuses all possible loops containing kernels."

hiker replied:
Done.

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/solution/fuse_loops.py:
```diff
@@ -0,0 +1,85 @@
+
+def trans(psyir):
+    '''
+    Take the supplied psyir object, and fuse all loops.
```
Please specify that this is written for a specific use case. You could challenge the user to make it general for bonus marks :-)

hiker replied:
Done. The challenge is in section  ... 2.18 or so

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/solution/openmp.py:
```diff
@@ -0,0 +1,70 @@
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Python script intended to be passed to PSyclone's generate()
+function via the -s option. It adds openmp for an MPI implementation.
```
Probably worth explaining what's special about it that it's "for an MPI implementation" - obviously there's no mention of MPI in the script and that could potentially be confusing.

hiker replied:
I've updated this.

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/time_step_alg_mod.x90:
```diff
@@ -0,0 +1,48 @@
+module time_step_alg_mod
```
Rather than have this file duplicated in `solution`, could we have a soft link to this file from that directory?

hiker replied:
We could. The reason why I prefer separate files is that it allows the trainees to play/modify the files, and still have access to a working version. Then again, it's all in git, so a user can always check.
For now, I've added this as a future todo to #3270.

arporter reviewed (changes requested):
Thanks Joerg, as you warned, this is big! I'm therefore wondering whether we could break it into chunks - I know github isn't going to be very happy by the time I've reviewed all 278 files.
So far, I've reviewed the changes outside the tutorial plus tutorial/training/2.10-GameOfLife-mpi. I'm therefore wondering whether we could have a PR for each of the separate 2.1x parts?

hiker commented on tutorial/training/Makefile:
```diff
@@ -0,0 +1,53 @@
+# -----------------------------------------------------------------------------
```
I've added them. Though I don't think the material is detailed enough for self-study (there's a lot of material in the actual slides)

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/Makefile:
```diff
@@ -0,0 +1,33 @@
+# sure we get a 2x2 domain decomposition, otherwise the job was not running properly
+# with MPI
+test: test-default
+	@bash -c "if [[ $$(make run | grep -c 'go_decompose: using grid of   2x  2') != 1 ]]; then \
```
I have done a lot of cleanup to the Makefiles since the review, so this line is gone now.

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/openmp.py:
```diff
@@ -0,0 +1,70 @@
+    omp_parallel_loop.omp_schedule = "dynamic"
+    for loop in schedule.walk(GOLoop):
+        if loop.loop_type == "outer":
+            omp_parallel_loop.apply(loop)
```
Done.

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/solution/openmp_combined.py:
```diff
@@ -0,0 +1,85 @@
+    # Look at the schedule before adding 'omp parallel':
+    print(schedule.view())
+
+    # Now add the OMP PARALLEL around all loops. In case of
```
Yes, it would - and it is also difficult to do in general ;)

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/solution/time_step_alg_mod.x90:
```diff
@@ -0,0 +1,48 @@
+module time_step_alg_mod
```
Done.

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,129 @@
+## MPI Support in the Infrastructure Library
+1. Parallel Initialisation
+The infrastructure can be compiled with MPI support. This library
+is then called `lib_dm_fd.a` (instead of `lib_df.a`). The API
```
Done.

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,129 @@
+## Adding Halo Exchanges
+The only required change is adding halo exchanges, since counting
+the number of neighbours means that updated information from neighbouring
+processes is required. And luckily, this error-prone step can be
```
Done.

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,129 @@
+To add MPI support, use the PSyclone command line flag `-dm` (for
+distributed memory):
+
+    psyclone -l output -dm -d /home/joerg/work/psyclone/tutorial/training/gol-lib -api gocean1.0 \
```
Oops.

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,129 @@
+You see that immediately before the kernel that counts the neighbours
+the halo exchange is automatically triggered.
+
+> Note that at this stage the infrastructure library (and therefore also
```
Done.

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,129 @@
+Fix the script.
+
+The reason for the error is the modified tree structure due to the added halo
+exchange. Fix the `openmp_combined.py` script to work here. Notice that later
```
Both done.

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/fuse_loops.py:
```diff
@@ -0,0 +1,85 @@
+    schedule = psyir.walk(InvokeSchedule)[0]
+
+    print(schedule.view())
+    # Inline all kernels to help gfortran with inlining.
```
Done.

hiker commented:
@arporter , I have addressed all changes. I also added a lot of additional tests, since I had at least two examples being broken that were not picked up by my testing - which was pretty bad in the actual training sessions. One issue is fixed already (and also properly tested here), one is still outstanding (#3252).

The fact that this tutorial contains additional important testing makes me push for getting this in sooner than later. Otherwise we risk a never-ending PR: as PSyclone gets updated and the training is not tested, there is a chance that something breaks the training, Then I need to debug and fix this (or get help to fix this), instead of these problems being flagged when the PSyclone improvements are added (and then being fixed by the original developer before the PR is merged).

Note that I have opened #3270 for additional training improvements (e.g. linking files instead of duplicating ... at least evaluating this option; adding syntax highlighting). And also #3208 to review the individual training sections. 

While I totally agree that this is not ideal, my suggestion would be that the review here focuses just on the testing side-of things, to get any issues there sorted out. No one has the time to proof-read all of this text material, and as mentioned, delays this PR just mean this will keep on being more work to keep this up-to-date. Yes, there is a risk that it might take a long time before this material (as in: the READMEs) gets properly updated, and I understand your reluctance.

I am open for any suggestion you have, I don't have a particular brilliant idea either :(

Am triggering the full CI as well.

arporter commented on tutorial/training/Makefile:
```diff
@@ -0,0 +1,51 @@
+# Author: J. Henrichs, Bureau of Meteorology
+
+# The transform target `test` runs tests on the gitlab CI (i.e. without
+# compilation), `test_run` will compile all binaries (as far as possible
```
Do you mean `compile` instead of `test_run` and then s/run/`test_run`/ on the following line?

hiker replied:
I've clean this up (including removing the 'compile' target from the top level makefiles ... most 'local' Makefiles have a compile for the convenience when running the training, but the high level ones should only use test or test_run)

arporter commented on tutorial/training/README.rst:
```diff
@@ -0,0 +1,90 @@
+The trainings material is split into four parts, and the files
+and documentation here are only the hands-on section. The full
+training is accompanied by a presentations
+(a slightly outdated version of the slides are available on the `wiki
```
Could you update this now that you've given the course?

hiker replied:
No.
:) Simple reason: that document was entirely manually created, I would have to manually replace all READMEs, all slides, in some cases code highlighting, ...
I have started to investigate options to create this automatically. My suggestion would be that I just upload the updated slide sets for now, and update the wiki to indicate that this version is outdated.

arporter replied:
Ah, I thought the slides were separate material to the training - I'd forgotten you created them. I like your suggestion since the training READMEs etc. are available in the repo anyway.

hiker replied:
Yes - that's going to be some fun (I have started some tests, e.g. converting README.md to pdf via a script). Next PR :)

arporter commented on tutorial/training/README.rst:
```diff
@@ -0,0 +1,90 @@
+introduction to PSyclone domain-specific-language features (DSL).
+It uses the easier ``gocean`` API to introduce the user to
+writing kernels, applying transformations, and writing
+transformation scripts. It is the longest session in this trainings
```
"training"

arporter commented on tutorial/training/README.rst:
```diff
@@ -0,0 +1,90 @@
+transformation scripts. It is the longest session in this trainings
+material and will take one full day.
+
+PSyclone Transformation API
```
I think this should be "Transforming existing Fortran code with PSyclone"

arporter commented on tutorial/training/README.rst:
```diff
@@ -0,0 +1,90 @@
+================
+The different training sections will provide makefiles for the users
+which will contain convenient shortcuts to avoid that a user has
+to type long command line options and file paths. But these targets
```
s/But/However,/

hiker replied:
Done.

arporter commented on tutorial/training/README.rst:
```diff
@@ -0,0 +1,90 @@
+The different training sections will provide makefiles for the users
+which will contain convenient shortcuts to avoid that a user has
+to type long command line options and file paths. But these targets
+are expected to fail in some cases, for example in the "PSyclone
```
s/for example in the/e.g. the/

arporter commented on tutorial/training/README.rst:
```diff
@@ -0,0 +1,90 @@
+are expected to fail in some cases, for example in the "PSyclone
+for LFRic Users" section contains examples that show how PSyclone
+will detect errors. In order to facilitate consistent testing
+across the whole trainings material, two additional targets
```
"training" and rather than say "additional targets have been added" (since it's not clear what they're additional to), you could have: "...material, the makefiles contain two special targets:"

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,137 @@
+processes is required. And happily, this error-prone step can be
+done automatically by PSyclone.
+
+PSyclone relies on the metadata provided for each kernel. Especially
```
You could provide a link to https://psyclone.readthedocs.io/en/latest/user_guide/gocean1p0.html#argument-metadata-meta-args here.

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/openmp.py:
```diff
@@ -0,0 +1,76 @@
+    # We know that there is only one schedule
+    schedule = psyir.walk(InvokeSchedule)[0]
+
+    # Call the existing fuse transformation, which will also inline.
```
"module inline"?

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/solution/fuse_loops.py:
```diff
@@ -0,0 +1,86 @@
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''
+Python script intended to be passed to PSyclone's generate()
```
Pls update (remove "generate()")

arporter commented on tutorial/training/gocean/2.10-GameOfLife-mpi/solution/openmp_combined.py:
```diff
@@ -0,0 +1,86 @@
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+'''
+Python script intended to be passed to PSyclone's generate()
```
Pls update.

arporter commented on tutorial/training/gocean/2.12-GameOfLife-extraction/Makefile:
```diff
@@ -0,0 +1,79 @@
+# to the local dl_no_mpi directory, and put this path first in the search list,
+# so PSyclone will only detect this version.
+# The include paths are:
+# - the gol-library directory (which is added automatically the Makefile, i.e.
```
"automatically by the"

arporter commented on tutorial/training/gocean/2.12-GameOfLife-extraction/Makefile:
```diff
@@ -0,0 +1,79 @@
+
+.precious: time_step_alg_mod.f90 time_step_alg_mod_psy.f90
+
+# When creating the driver, we need to specify the paths to all module directories
```
You have this block comment twice?

arporter commented on tutorial/training/gocean/2.12-GameOfLife-profiling/dl_no_mpi/parallel_utils_stub_mod.f90:
```diff
@@ -0,0 +1,164 @@
+!------------------------------------------------------------------------------
```
Sym link again?

hiker replied:
This file is not required anymore (in the past I had all psydata material in one directory, with different Makefiles. That was just way too confusing, so I split this, and missed removing this file). All fixed now.

arporter commented on tutorial/training/gocean/2.12-GameOfLife-value-range-check/compute_born_mod.f90:
```diff
@@ -0,0 +1,54 @@
+        ! A new cell is born in an empty location if
+        ! it has exactly three neighbours.
+        if (current(i, j) == 0.0 .and. neighbours(i,j) == 3.0) then
+            born(i, j) = 1.0
```
Why do you assign to `born(i,j)` twice?

arporter commented on tutorial/training/gocean/2.12-GameOfLife-value-range-check/dl_no_mpi/parallel_utils_stub_mod.f90:
```diff
@@ -0,0 +1,164 @@
+!------------------------------------------------------------------------------
```
sym link?

arporter commented on tutorial/training/gocean/2.12-GameOfLife-value-range-check/value_range_check.py:
```diff
@@ -0,0 +1,61 @@
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Python script intended to be passed to PSyclone's generate()
```
Pls update.

arporter commented on tutorial/training/gocean/2.14-GameOfLife-openacc/openacc.py:
```diff
@@ -0,0 +1,91 @@
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Python script intended to be passed to PSyclone's generate()
```
Pls update.

arporter commented on tutorial/training/gocean/2.14-GameOfLife-openacc/Makefile:
```diff
@@ -0,0 +1,25 @@
+
+OBJ = output_field_mod.o time_step_alg_mod.o time_step_alg_mod_psy.o
+
+# Use nvfortran by default
```
Perhaps specify which memory mode is used (or at least, that managed memory isn't used).

arporter commented on tutorial/training/gocean/2.14-GameOfLife-openacc/solution/Makefile:
```diff
@@ -0,0 +1,26 @@
+
+# We can't simply use the default test, since it will require to recompile
+# the infrastructure files. So only test the transformation here.
+# If compilation is to be tested, use `make test-default`
```
But the following target depends on `test-default` - I must be missing something?

hiker replied:
That comment did not match the code anymore :) Fixed.

arporter commented on tutorial/training/gocean/2.16-GameOfLife-omp-offload/fuse_loops.py:
```diff
@@ -0,0 +1,98 @@
+
+def trans(psyir):
+    '''
+    Take the supplied psyir object, and fuse the first two loops
```
It does more than that.

arporter commented on tutorial/training/gocean/2.16-GameOfLife-omp-offload/fuse_loops.py:
```diff
@@ -0,0 +1,98 @@
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Python script intended to be passed to PSyclone's generate()
```
Please update and correct: s/adds kernel fusion code to/fuses all possible loops in/

hiker replied:
Done.

arporter commented on tutorial/training/gocean/2.16-GameOfLife-omp-offload/output_field_mod.f90:
```diff
@@ -0,0 +1,33 @@
+module output_field_mod
```
Docstring please.

arporter commented on tutorial/training/gocean/2.16-GameOfLife-omp-offload/solution/fuse_loops.py:
```diff
@@ -0,0 +1,87 @@
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Python script intended to be passed to PSyclone's generate()
+function via the -s option. It adds kernel fuseion code to
```
Please correct.

arporter commented on tutorial/training/gocean/2.16-GameOfLife-omp-offload/solution/omp_offload.py:
```diff
@@ -0,0 +1,89 @@
+
+def trans(psyir):
+    '''
+    Take the supplied psy object, and fuse the first two loops
```
Please correct.

arporter commented on tutorial/training/gocean/2.16-GameOfLife-omp-offload/solution/omp_offload.py:
```diff
@@ -0,0 +1,89 @@
+        for loop in subroutine.walk(Loop):
+            if loop.loop_type == "outer":
+                loop_offloading.apply(
+                    loop, options={"independent": True})
```
Pls add a comment about why this option is required.

arporter commented on tutorial/training/gocean/2.18-GameOfLife-IntelligentScripts/solution/fuse_loops_index.py:
```diff
@@ -0,0 +1,132 @@
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Python script intended to be passed to PSyclone's generate()
```
Please update.

arporter commented on tutorial/training/gocean/2.18-GameOfLife-IntelligentScripts/solution/fuse_loops_index.py:
```diff
@@ -0,0 +1,132 @@
+def trans(psyir):
+    '''
+    Take the supplied psy object, and fuse loops as much as possible.
+    :param psyir: the PSyIR of the PSy-layer.
```
Blank line pls.

arporter commented on tutorial/training/gocean/2.18-GameOfLife-IntelligentScripts/solution/fuse_loops_siblings.py:
```diff
@@ -0,0 +1,142 @@
+
+def apply_all(node_list, transform):
+    '''This subroutine applies the specified transformation, which takes
+    two consecutive nodes (e.g. loop fusion), and applies it to a block
```
This needs re-wording.

arporter commented on tutorial/training/gocean/2.2-GameOfLife/Makefile:
```diff
@@ -0,0 +1,21 @@
+	  output_field_mod.f90
+OBJ = $(SRC:%.f90=%.o)
+
+test transform:
```
So what does this Makefile do? Please add a comment at the top perhaps?

arporter commented on tutorial/training/gocean/2.2-GameOfLife/README.md:
```diff
@@ -0,0 +1,184 @@
+# Introduction to ``dl_esm_inf``
+
+This tutorial introduces the infrastructure library
+``dl_esm_inf``, which is used by the PSyclone
```
Worth linking to the User Guide documentation on this here (https://psyclone.readthedocs.io/en/latest/user_guide/gocean1p0.html#the-gocean-infrastructure-library-dl-esm-inf)

arporter commented on tutorial/training/gocean/2.2-GameOfLife/read_config_mod.f90:
```diff
@@ -0,0 +1,98 @@
+module read_config_mod
+
```
Doc string please as this does more than just "reading"

hiker replied:
I've updated all doc strings of all Fortran files that I've found. I won't comment on the other comments re this from you, they should all be ok now.

arporter commented on tutorial/training/gocean/2.6-GameOfLife-fuse/solution/fuse_loops.py:
```diff
@@ -0,0 +1,99 @@
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+
+'''Python script intended to be passed to PSyclone's generate()
```
Pls update.

arporter commented on tutorial/training/gocean/2.6-GameOfLife-fuse/solution/fuse_loops.py:
```diff
@@ -0,0 +1,99 @@
+
+def trans(psyir: FileContainer) -> None:
+    '''
+    Take the supplied psyir object, and fuse the first two loops
```
Pls update.

arporter commented on tutorial/training/gocean/2.4-GameOfLife-psyclone/solution/read_config_mod.f90:
```diff
@@ -0,0 +1,97 @@
+
+    line_count = 0
+
+    do
```
Comment please.

hiker commented on tutorial/training/README.rst:
```diff
@@ -0,0 +1,90 @@
+PSyclone Training
```
Done.

hiker commented on tutorial/training/gocean/2.12-GameOfLife-value-range-check/compute_born_mod.f90:
```diff
@@ -0,0 +1,54 @@
+        born(i, j) = 0.0
+        ! A new cell is born in an empty location if
+        ! it has exactly three neighbours.
+        if (current(i, j) == 0.0 .and. neighbours(i,j) == 3.0) then
```
Indeed, it doesn't. Plus, the integer values 0 to 8 (which are all that are used here) can be represented exactly, so this is not  a problem here (though of course in general you are right).

hiker commented on tutorial/training/gocean/2.12-GameOfLife-value-range-check/solution/compute_born_mod.f90:
```diff
@@ -0,0 +1,54 @@
+
+    !> @brief Computes which empty fields will get a new cell.
+    !>
+    !> This subroutine computes a field 'born' which has the value '1'
```
Done, or something similar at least :)

hiker commented on tutorial/training/gocean/2.2-GameOfLife/solution/time_step_mod.f90:
```diff
@@ -0,0 +1,48 @@
+        die        = r2d_field(grid, GO_T_POINTS)
+        born       = r2d_field(grid, GO_T_POINTS)
+
+        if(time_steps<=20) call output_field(current)
```
I've added a comment (to print the initial state, so you can even see what the first time step does).

hiker commented on tutorial/training/README.rst:
```diff
@@ -0,0 +1,90 @@
+more open ended, and trainees are encouraged to go back to these
+examples and implement better solutions later.
+
+The trainings material is split into four parts, and the files
```
Done.

hiker commented on tutorial/training/README.rst:
```diff
@@ -0,0 +1,90 @@
+training is accompanied by a presentations
+(a slightly outdated version of the slides are available on the `wiki
+<https://github.com/stfc/PSyclone/wiki/Presentations>`_). It
+contains the following session (in order in which they are presented):
```
Done.

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,137 @@
+process needs to read in the global field), but given the very
+simple and short input file size this is acceptable for this application.
+
+3. Gathering Local Data
```
Done (though I think it's not entirely wrong: you collect the local data on each PE into a global array on the master).

hiker commented on tutorial/training/gocean/2.10-GameOfLife-mpi/README.md:
```diff
@@ -0,0 +1,137 @@
+## Adding Halo Exchanges
+The only required change is adding halo exchanges, since counting
+the number of neighbours means that updated information from neighbouring
+processes is required. And happily, this error-prone step can be
```
Done.

hiker commented on tutorial/training/gocean/2.12-GameOfLife-value-range-check/solution/compute_born_mod.f90:
```diff
@@ -0,0 +1,54 @@
+        ! A new cell is born in an empty location if
+        ! it has exactly three neighbours.
+        if (current(i, j) == 0.0 .and. neighbours(i,j) == 3.0) then
+            born(i, j) = 1.0
```
Removed.

hiker commented on tutorial/training/gocean/2.14-GameOfLife-openacc/output_field_mod.f90:
```diff
@@ -0,0 +1,34 @@
+        real(go_wp), dimension(:,:), allocatable :: global_data
+        integer                                  :: j
+
+!$acc update self(field%data)
```
Done.

hiker commented on tutorial/training/gocean/2.16-GameOfLife-omp-offload/Makefile:
```diff
@@ -0,0 +1,23 @@
+
+OBJ = output_field_mod.o time_step_alg_mod.o time_step_alg_mod_psy.o
+
+# Use nvfortran by default
```
Done.

hiker commented on tutorial/training/gocean/2.16-GameOfLife-omp-offload/solution/output_field_mod.f90:
```diff
@@ -0,0 +1,34 @@
+        real(go_wp), dimension(:,:), allocatable :: global_data
+        integer                                  :: j
+
+!$acc update self(field%data)
```
I can't run the managed memory version, but you did for me (quite some time ago), and confirmed that it works. My understanding is that this is not required in managed memory(?), so that's why there is no omp equivalent statement here. I can run openacc, but I do need that statement.

hiker commented on tutorial/training/gocean/2.18-GameOfLife-IntelligentScripts/Makefile:
```diff
@@ -0,0 +1,22 @@
+	$(PSYCLONE) -s ./fuse_loops.py -oalg $*.f90 -opsy $*_psy.f90 $<
+
+test:
+	@echo Ignoring test due to known bug.
```
Outdated and removed (I fixed the solution Makefile, but missed this one)

hiker commented on tutorial/training/gocean/2.2-GameOfLife/solution/Makefile:
```diff
@@ -0,0 +1,22 @@
+include ../../common.mk
+
+test transform:
+	# Nothing to do here, testing requires compilation
```
Done

hiker commented on tutorial/training/gocean/2.4-GameOfLife-psyclone/combine_mod.f90:
```diff
@@ -0,0 +1,43 @@
+    !>
+    !> @param[in]    i, j Coordinates of the cell to update.
+    !> @param[inout] current The current state that will be updated.
+    !> @param[in]    die     The field with 1 iff the cell dies.
```
Hmm - I thought 'iff' = 'if and only if' is common. But I agree (even if it were common), it's not entirely correct (since iff only applies to logical statements, and we have reals :) ). So, I have changed all `iff` everywhere!

hiker commented on tutorial/training/gocean/2.6-GameOfLife-fuse/README.md:
```diff
@@ -0,0 +1,252 @@
+The original design of splitting the code into several
+small kernels is typically not ideal from a performance
+point of view, since hardly any data is being reused and
+data needs to be loaded from memory, instead from cache.
```
Done.

hiker commented:
@arporter I have a problem - turns out the OpenMP offloading does not work (anymore??). Due to a bug in the makefile, it actually hasn't applied the offloading script at all in my recent tests (I am pretty certain it did originally, but I did some refactoring/ simplifications of the Makefile, which I believe introduced this bug recently). Now I've fixed the Makefile, it aborts with:
```
make -C 2.16-GameOfLife-omp-offload/solution test_run
...
make --no-print-directory  run | tail -n 12 | diff -b - /archive/psyclone-tests/action-runner-software/actions-runner/_work/PSyclone-mirror/PSyclone-mirror/tutorial/training/gocean/gol-lib/glider.correct

libgomp: pointer target not mapped for attach
```
(see https://github.com/stfc/PSyclone-mirror/actions/runs/20805411529/job/59758702708)

Googling indicates that there might be a mapping missing (which we don't support afaik), or additional compiler options(?). I don't know what to do. I don't have (easy) access to hardware that would support managed memory, and I am not even certain what the right solution would be in this case :( Can you (or Sergi?) help? I assume you can log onto the machine and try running it there?

hiker commented:
As discussed, I have added a `USE_GPU=yes/no` variable. If this is set to `no`, the GPU examples will be compiled, but not executed (even without openacc/opnmp support they can be compiled, which is at least a bit better than just checking if the transformations are done).

I only triggered the compilation IT (to speed things up), since I expect there will be another round of changes, and no code here should affect anything else outside of training.

arporter commented on tutorial/training/gocean/2.8-GameOfLife-openmp/solution/Makefile:
```diff
@@ -0,0 +1,34 @@
+	cat time_step_alg_mod_psy.f90 | grep -iq "omp *taskloop"
+	@# The fuse_loop scripts are used in openmp and openmp_combined
+
+test_run: #test
```
Pls remove the commented part.

arporter commented on tutorial/training/lfric/4.2-openmp/Makefile:
```diff
@@ -0,0 +1,95 @@
+API = lfric
+include ../common.mk
+
+# Path of the included simplified LFRic infrastructure files:
```
rm comment? (Radical, I know)

arporter commented on tutorial/training/lfric/4.4-colouring-openmp/omp_colour_transformation.py:
```diff
@@ -0,0 +1,71 @@
+
+
+def trans(psy):
+    '''PSyclone transformation script for the dynamo0p3 api to apply
```
s/dynamo0p3/LFRic/ :-)

arporter replied:
Also, please update this to use `psyir` rather than `psy` and thus remove the `psy.invokes.invoke_list` usage.

arporter commented on tutorial/training/lfric/4.4-colouring-openmp/omp_transformation.py:
```diff
@@ -0,0 +1,55 @@
+
+
+def trans(psy):
+    '''PSyclone transformation script for the dynamo0p3 api to apply
```
Pls update and change psy to psyir too.

arporter commented on tutorial/training/lfric/4.4-colouring-openmp/solution/summation_w0_to_w0_kernel_mod.f90:
```diff
@@ -0,0 +1,83 @@
+!$            " using indices ", map_w3(1), " - ", map_w3(1)+nlayers-1
+    do k=0, nlayers-1
+      do i=1, ndf_w3
+        fld1(map_w3(i)+k) = fld1(map_w3(i)+k) + fld2(map_w3(i)+k)
```
I've not looked at the tutorial text but hopefully you point out that operations like this are actually supported natively as Built-ins?

hiker replied:
No ... because it is not. That example is first of all broken as I just realise (I provided a field on W3 as a field on W0 argument, which makes no sense ... :( ).

Then, this code actually adds elements several time: a vertex in the 'middle' of the grid (i.e. created by or surrounded by 8 3D-FE cells) will add the value 8 times (since the summation will be executed 8 times, one for each 3D element when it is computed). Since the fields are initialised with 0 (output) and 1 (input), this function actually kind of counts by how many 3D elements it is created: a vertex at the corner of the grid will have 1, a vertex at the edge of the whole grid will have 2, .... Overall, you get values between 1 and 8.

The builtin will (I assume) just add the elements one (i.e 0+1, i.e. the result will be 1 for all vertices).

That totally confused me the very first time I tried to execute a kernel and I tried to understand why my numbers did not add up (pun intended) :)

I feel we should have a better example, one that might be more ... sensible? Dave told me that this example makes absolutely no sense from a FE point of view :)

For now, I'll fix up this example to properly take two W0 fields.

arporter commented on tutorial/training/lfric_users/6_missing_parameter/main_alg.x90:
```diff
@@ -0,0 +1,129 @@
+                 setval_c( field_3,     0.0_r_def ), &
+                 setval_c( field_0,     1.0_r_def )  &
+                 )
+    call invoke( name = 'summation', summation_w0_to_w3_kernel_type(field_3) )
```
Comment please on the missing parameter.

hiker replied:
I first added the comment, then removed it again. Yes, the idea is that the user will find this out themselves.

arporter commented on tutorial/training/lfric_users/6_missing_parameter/summation_w0_to_w3_kernel_mod.f90:
```diff
@@ -0,0 +1,86 @@
+contains
+
+    ! This kernel adds all 8 values from the field on W0 to the corresponding
+    ! element in W3. It assumes lowest order finite elements.
```
Assumption?

hiker replied:
Removed.

arporter commented on tutorial/training/lfric_users/common.mk:
```diff
@@ -0,0 +1,40 @@
+# in this list is the current file.
+this_file := $(abspath $(lastword $(MAKEFILE_LIST)))
+
+# PSyclone directory is up two from this file
```
s/two/three/

hiker replied:
Fixed.

arporter commented on tutorial/training/lfric_users/fem-grid-summation.png:
Sym link to a common location?

arporter commented on tutorial/training/lfric_users/fem-grid-summation.svg:
Do you need both the svg and the png?

arporter commented on tutorial/training/lfric_users/fem-grid.svg:
svg as well as png?

arporter commented on tutorial/training/transformation/3.2-openmp/Makefile:
```diff
@@ -0,0 +1,38 @@
+# ------------------------
+.precious: $(SRC_X90:%.x90=%.f90)
+
+# This will need an additional flag
```
Not sure what this means?

arporter commented on tutorial/training/transformation/3.2-openmp/compute_die_mod.x90:
```diff
@@ -0,0 +1,51 @@
+                    ! A cell dies by underpopulation if it has less than 2
+                    ! neighbours or by overpopulation if it has more than 3
+                    if (neighbours(i, j) < 2.0 .or. &
+                        neighbours(i, j) > 3.0       ) die(i, j) = 1.0
```
.f90?

hiker replied:
Done.

arporter commented on tutorial/training/transformation/3.2-openmp/count_neighbours_mod.x90:
```diff
@@ -0,0 +1,43 @@
+        do j=ystart, ystop
+            do i=xstart, xstop
+                neighbours(i, j) = &
+                      current(i-1, j-1) + current(i, j-1) + current(i+1, j-1) &
```
.f90?

arporter commented on tutorial/training/transformation/3.2-openmp/solution/time_step_mod.x90:
Could be a .f90?

hiker replied:
Done.

arporter commented on tutorial/training/transformation/3.4-tiling/solution/tiling_with_openmp.py:
```diff
@@ -0,0 +1,160 @@
+            except TransformationError as err:
+                print(f"Cannot move code {child}: {err.value}")
+
+    const_mapping = {}
```
I like this. Do you need to do this to encourage loop fusion? We do have https://psyclone.readthedocs.io/en/latest/_autogenerated/autoapi/psyclone.psyir.transformations.replace_reference_by_literal_trans.html#module-psyclone.psyir.transformations.replace_reference_by_literal_trans but that only works when the RHS is a Literal.

hiker replied:
Yes, without this loop fusion aborts (I believe that is explained in the README as well). And I tried this, it indeed doesn't work, since we don't have constants.

arporter commented on tutorial/training/transformation/3.4-tiling/time_step_mod.x90:
```diff
@@ -0,0 +1,50 @@
+        ! 3. Compute dying cells.
+        ! 4. Combine the results from 2. and 3. with current state
+        do time =1, time_steps
+            call count_neighbours(neighbours, current)
```
Just a .f90?

hiker replied:
Done.

arporter commented on tutorial/training/transformation/3.6-sympy/dataflow.svg:
Nice :-)

hiker replied:
Thanks!

arporter commented on tutorial/training/transformation/3.6-sympy/solution/dataflow.py:
```diff
@@ -0,0 +1,137 @@
+# Create the PSyIR, and get the variable access information:
+reader = FortranReader()
+psyir = reader.psyir_from_source(code)
+routine = psyir.find_routine_psyir("foo")
```
You don't seem to use `routine` but perhaps you should s/psyir.children[0]/routine/ below?

arporter commented on tutorial/training/transformation/3.8-code-creation/solution/combine_mod.x90:
```diff
@@ -0,0 +1,41 @@
+
+    subroutine combine(current, die, born)
+
+        implicit none
```
.f90?

arporter reviewed (changes requested):
Made it all the way through! I like your various examples - they are impressive.
Mostly it's just a bit more tidying of docstrings.
I have to admit that I'm not that keen on the use of the `.x90` extension on files that are just standard `.f90`. This could cause confusion with the the Met Office approach where only files written using the DSL have the `.x90` extension. I can see that it perhaps makes it easier to automatically run PSyclone on the necessary files but you could just name them in the Makefile.

hiker commented on tutorial/training/gocean/2.8-GameOfLife-openmp/solution/Makefile:
```diff
@@ -0,0 +1,34 @@
+	SCRIPT=./openmp_combined.py $(MAKE) test_run-default
+	OMP_NUM_THREADS=2 $(MAKE) run | grep  "Using *2 *threads" > /dev/null
+
+	@# Taskloop is not supported by nvfortran, so disable this test for now:
```
I re-checked,:
```
NVFORTRAN-F-1235-The OpenMP 'taskloop' construct is not supported. (time_step_alg_mod_psy.f90: 20)
NVFORTRAN/x86-64 Linux 25.7-0: compilation aborted
```

arporter replied:
Thanks for checking and adding the compiler version.

hiker commented on tutorial/training/gocean/2.18-GameOfLife-IntelligentScripts/fuse_loops.py:
```diff
@@ -58,9 +58,9 @@ def trans(psyir):
     schedule = psyir.walk(InvokeSchedule)[0]
 
     # Inline all kernels to help gfortran with inlining.
```
Done.

hiker commented on tutorial/training/gocean/gol-lib/count_neighbours_mod.f90:
```diff
@@ -32,8 +32,10 @@ module count_neighbours_mod
     !> @param[out] c    The current state that will be updated.
-    !> @param[in]  die  The field with 1 iff the cell dies.
-    !> @param[in]  born The field with 1 iff a new cell is born.
+    !> @param[in]  die  The field with 1 if the cell dies,
```
Fixed (I hope in all files across all training *sigh*)

hiker commented on tutorial/training/lfric/4.4-colouring-openmp/fem-grid.png:
I re-did the whole image handling. There is now `tutorial/training/images`, which contains the svg files (editable), and the png files (used in READMEs). While it is possible to use svg in markdown, I could not get them to be scaled appropriately (i.e. not too big). I even tried using `<img...` tags directly (as this is the official solution), but it just didn't work for me, any .svg file stayed too big. So I just use the png everywhere. All READMEs now reference the files in this image directory.

hiker commented on tutorial/training/lfric/4.4-colouring-openmp/solution/update_w0_with_w0_kernel_mod.f90:
```diff
@@ -0,0 +1,84 @@
+
+contains
+
+    ! This kernel adds all 8 values from the field on W0 to the corresponding
```
I've removed this file.

hiker commented on tutorial/training/lfric/common.mk:
```diff
@@ -0,0 +1,68 @@
+# arguments passed to the first invocation of Make. The last entry
+# in this list is the current file.
+this_file := $(abspath $(lastword $(MAKEFILE_LIST)))
+# PSyclone directory is up two from this file
```
Done

hiker commented on tutorial/training/lfric/common.mk:
```diff
@@ -0,0 +1,68 @@
+clean-default:
+	rm -f *.o $(EXE) *.mod time_step_alg_mod.f90 time_step_alg_mod_psy.f90
+
+# The target allclean cleans all also all remote dependencies
```
No idea what I meant there :) Reworded.

hiker commented on tutorial/training/lfric_users/3_applying_openmp/summation_w0_to_w3_kernel_mod.f90:
```diff
@@ -0,0 +1,89 @@
+
+      integer(kind=i_def)                                 :: i, k
+
+!$    print *,"Kernel executed by thread ", omp_get_thread_num(), " of ", omp_get_max_threads(), &
```
Removed.

hiker commented on tutorial/training/lfric_users/4_mpi_and_openmp/Makefile:
```diff
@@ -0,0 +1,89 @@
+
+run: compile
+	@echo =============================================
+	@echo Warning: this example does not run as MPI job
```
Done.

hiker commented on tutorial/training/lfric_users/5_alg_layer_error/summation_w0_to_w3_kernel_mod.f90:
```diff
@@ -0,0 +1,86 @@
+contains
+
+    ! This kernel adds all 8 values from the field on W0 to the corresponding
+    ! element in W3. It assumes lowest order finite elements.
```
Removed.

hiker commented on tutorial/training/lfric_users/7_invalid_openmp/omp_colour_transformation.py:
```diff
@@ -0,0 +1,71 @@
+
+
+def trans(psyir: Node):
+    '''PSyclone transformation script for the dynamo0p3 api to apply
```
Done.

hiker commented on tutorial/training/lfric_users/7_invalid_openmp/omp_transformation.py:
```diff
@@ -0,0 +1,56 @@
+
+
+def trans(psyir: Node):
+    '''PSyclone transformation script for the dynamo0p3 api to apply
```
Done.

hiker commented on tutorial/training/lfric_users/8_incorrect_naming/main_alg.x90:
```diff
@@ -0,0 +1,129 @@
+
+    field_0_proxy = field_0%get_proxy()
+
+    print *,"minmax of field1", field_0_proxy % get_min(), field_0_proxy % get_max()!
```
Oh, and they are everywhere. All removed :)

hiker commented on tutorial/training/lfric_users/fem-grid.png:
Replaced reference in README to use ../images

hiker commented on tutorial/training/transformation/3.2-openmp/omp_parallel.py:
```diff
@@ -0,0 +1,75 @@
+    for loop in psyir.walk(Loop):
+        if loop.loop_type == "lat":
+            # Apply transformation. Note that you need to specify
+            # "--backend disable-validation" on the PSyclone command line,
```
Yep :)

hiker commented on tutorial/training/transformation/3.2-openmp/solution/count_neighbours_mod.f90:
```diff
@@ -0,0 +1,43 @@
+            do i=xstart, xstop
+                neighbours(i, j) = &
+                      current(i-1, j-1) + current(i, j-1) + current(i+1, j-1) &
+                    + current(i-1, j  )                   + current(i+1, j  ) &
```
Done.

hiker commented on tutorial/training/transformation/3.2-openmp/solution/omp_trans.py:
```diff
@@ -0,0 +1,68 @@
+    '''
+    # Get the transformation we will apply
+    ompt = OMPParallelLoopTrans()
+    # Apply it to each loop over levels containing a kernel
```
Done.

hiker commented on tutorial/training/transformation/3.4-tiling/tiling_with_openmp.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def trans(psyir):
+    '''A complex program that inline all loops, moves the scalar assignment to
```
Done.

hiker commented on tutorial/training/transformation/common.mk:
```diff
@@ -0,0 +1,50 @@
+# arguments passed to the first invocation of Make. The last entry
+# in this list is the current file.
+this_file := $(abspath $(lastword $(MAKEFILE_LIST)))
+# PSyclone directory is up two from this file
```
Done.

hiker commented:
The documentation failure is caused by the link `https://gtr.ukri.org/projects?ref=NE%2FL01209X%2F1`, so not this PR's problem.

The IT failed in the lfric intregration test. It looks like an nvidia compiler crash:
```
nvfortran-Fatal-/archive/psyclone-spack/psyclone-spack-Jun25/spack-repo/opt/spack/linux-rhel10-zen5/gcc-14.2.1/nvhpc-25.1-e6n46oh4nhjizqwqswfzzfirgardslid/Linux_x86_64/25.1/compilers/bin/tools/fort1 TERMINATED by signal 11
```
(line 40091) in the lfric gungho with openmp offload section. Again, I can't see how this PR should cause this? And I can't figure out which file might be causing this?

arporter commented:
> (line 40091) in the lfric gungho with openmp offload section. Again, I can't see how this PR should cause this? And I can't figure out which file might be causing this?

Don't worry. We see this sometimes. Re-running the job almost always fixes it!

arporter commented on tutorial/training/gocean/2.12-GameOfLife-value-range-check/solution/value_range_check.py:
```diff
@@ -39,15 +39,15 @@
+def trans(psyir: FileContainer) -> None:
     '''
-    Take the supplied psy object, and apply 'omp parallel do' to all loops.
+    Take the supplied PSyOR object, and apply 'omp parallel do' to all loops.
```
"PSyIR"

arporter commented on tutorial/training/lfric/4.2-openmp/openmp.py:
```diff
@@ -44,22 +44,24 @@
-    :type psy: :py:class:`psyclone.psyGen.PSy`
+    Take the supplied psyir object, apply module inlining and add generic
+    OpenMP parallelisation to the code. Also check if loop fusion can
+    be applied. Also check if loop fusion can
```
Duplicated last sentence.
